walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,117 @@
+            int currentBusyStart = busyTimes.get(i).start();
+            int currentBusyEnd = busyTimes.get(i).end();
+            int nextBusyStart = busyTimes.get(i+1).start();
+            int nextBusyEnd = busyTimes.get(i+1).end();
+            int adjustedNewStart;
+            int adjustedNewEnd;
+
+            //if the current busy time contains the new event, or they are equal, do not add the new event to the arraylist as it already exists
+            /*****/  //current busy block
+             /**/    //new event
+            if (busyTimes.get(i).contains(newEvent) || busyTimes.get(i).equals(newEvent)){
+                break; 
+            }
+            //if the new event starts during the current busy time and ends before the next busy time
+            /*******/        /**/       //busy blocks
+                 /******/               //new event
+            else if (TimeRange.contains(busyTimes.get(i), newEventStart) && newEventEnd <= nextBusyStart) {
+                adjustedNewStart = currentBusyEnd;
+                adjustedNewEnd = newEventEnd;
+            }
+            //if the new event's start is during the current busy time and it's end is in the next busy time, fill in the time between the busy times
+            /****/   /****/     //busy blocks
+              /********/        //new event
+            else if (TimeRange.contains(busyTimes.get(i), newEventStart) && TimeRange.contains(busyTimes.get(i+1), newEventEnd)) {
+                adjustedNewStart = currentBusyEnd;
+                adjustedNewEnd = nextBusyStart;
+            }
+            //if the new event is strictly between the current and next busy times, add it as is
+            /***/        /***/      //busy blocks
+                   /**/             //new event
+            else if (newEventStart >= currentBusyEnd && newEventEnd <= nextBusyStart) {
+                adjustedNewStart = newEventStart;
+                adjustedNewEnd = newEventEnd;
+            }
+            //if it's start time is after the current busy block's end time and it's end time is during the next busy time
+            /***/        /***/      //busy blocks
+                    /*****/         //new event            
+            else if (newEventStart >= currentBusyEnd && newEventStart < nextBusyStart && TimeRange.contains(busyTimes.get(i+1), newEventEnd)) {
+                adjustedNewStart = newEventStart;
+                adjustedNewEnd = nextBusyStart;
+            }
+            //the new event does not start during the current busy block or before the next busy block, so move on to next busy block
+            else {
+                continue;
+            }
+
+            //add the new busy time with the adjusted start and duration
+            busyTimes.add(TimeRange.fromStartEnd(adjustedNewStart, adjustedNewEnd, false));
+            //sort the list in chronological order by event start time
+            Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
```

Comment:
sorting is costly! I'd rather you add events at the end, iterate over all of them each time and then sort everything at the end. As I pointed out above there are edge cases in which checking the two neighbors is not sufficient anyway. Alternatively you can try to use a TreeSet which sorts on insert.